File: Linked List/doubly_linked_list.java
/*A Doubly Linked List is a data structure in which each node contains a reference to the previous node and the next node in the sequence. It extends the functionality of a Singly Linked List by allowing traversal in both forward and backward directions.The first node of the list is referred to as the head, and the last node is referred to as the tail. In this implementation, the Node class represents a node in the doubly linked list. Each node has a data field to store the value, and prev and next fields to maintain references to the previous and next nodes in the list.*/

class Node {
    int data;
    Node prev;
    Node next;
    
    public Node(int data) {
        this.data = data;
        this.prev = null;
        this.next = null;
    }
}

class DoublyLinkedList {
    Node head;
    
    public DoublyLinkedList() {
        this.head = null;
    }
    
    // Insert a new node at the end of the list
    public void insert(int data) {
        Node newNode = new Node(data);
        
        if (head == null) {
            // If the list is empty, make the new node the head
            head = newNode;
        } else {
            Node current = head;
            
            // Traverse to the end of the list
            while (current.next != null) {
                current = current.next;
            }
            
            // Link the new node to the last node
            current.next = newNode;
            newNode.prev = current;
        }
    }
    
    // Delete a node with the given data value
    public void delete(int data) {
        Node current = head;
        
        while (current != null) {
            if (current.data == data) {
                if (current.prev != null) {
                    // If the node to be deleted is not the first node
                    
                    // Update the previous node's next reference
                    current.prev.next = current.next;
                } else {
                    // If the node to be deleted is the first node
                    
                    // Update the head reference to the next node
                    head = current.next;
                }
                
                if (current.next != null) {
                    // If the node to be deleted is not the last node
                    
                    // Update the next node's previous reference
                    current.next.prev = current.prev;
                }
                
                break;
            }
            
            current = current.next;
        }
    }
    
    // Display the elements of the list
    public void display() {
        Node current = head;
        
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }
        
        System.out.println();
    }
}

public class Main {
    public static void main(String[] args) {
        DoublyLinkedList list = new DoublyLinkedList();
        
        list.insert(10);
        list.insert(20);
        list.insert(30);
        list.insert(40);
        
        list.display(); // Output: 10 20 30 40
        
        list.delete(20);
        list.delete(40);
        
        list.display(); // Output: 10 30
    }
}

/* Time complexity: O(n) for insertions and deletions, and O(1) for display.
Space complexity: O(1) for all operations */

File: Linked List/doubly_linked_list.cpp
/* Name : Aneesh
Github username : 007aneesh
Repository name : data-structures-and-algorithms

Problem : Doubly linked list in c++
Issue Number : #975
Problem statement : Given an integer n representing the length of linked list and you have to print linked list from start and end.

Sample testcases: 

Testcase 1 --> 

Input: 
5
1 2 3 4 5

Output:
Printing forwardly................................
1
2
3
4
5
Printing backwardly................................
5
4
3
2
1


Testcase 2 -->
Input: 
3
10 20 30

Output:
Printing forwardly................................
10
20
30
Printing backwardly................................
30
20
10



Time Complexity = O(n)
Space Complexity = O(n)


Explanation:
In the main function, the program first takes input n,
the number of nodes to be inserted in the linked list.
Then, in a loop, it takes n values as input and creates new nodes with those values.
The nodes are then linked together to form a doubly linked list. 
Finally, it calls the PrintForward and PrintBackward functions
to print the elements of the list in forward and backward order, respectively.
*/

// ----------------------------------------------------------------------------- code begins now!

#include <iostream>
using namespace std;

// Define a Node class
class Node {
public:
    int data;
    Node* next;
    Node* prev;
    Node(int d) {
        this->data = d;
        next = NULL;
        prev = NULL;
    }
};

// Function to print the linked list forward
void PrintForward(Node* head) {
    Node* traverse = head;
    while (traverse != NULL) {
        cout << traverse->data << endl;
        traverse = traverse->next;
    }
}

// Function to print the linked list backward
void PrintBackward(Node* tail) {
    Node* traverse = tail;
    while (tail != NULL) {
        cout << traverse->data << endl;
        traverse = traverse->prev;
    }
}

int main() {
    int n, value;
    cin>>n; // Read the number of nodes to be created
    Node *head = nullptr; // Initialize head pointer
    Node *tail = nullptr; // Initialize tail pointer

    // Read 'n' values and create a doubly linked list
    for (int i = 0; i < n; i++) {
        cin >> value; // Read the value for the current node
        Node* newNode = new Node(value);  // Create a new node with the value
        
        // If the list is empty, set the head and tail to the new node
        if (head == NULL) {
            head = newNode;
            tail = newNode;
        } else {
            // Add the new node to the end of the list
            newNode->next = nullptr;
            newNode->prev = tail;
            tail->next = newNode;
            tail = newNode;
        }
    }

    // Print the linked list forward
    cout << "Printing forwardly................................" << endl;
    PrintForward(head);

    // Print the linked list backward
    cout << "Printing backwardly................................" << endl;
    PrintBackward(tail);

    return 0;
}
